<p style="font-family: Arial; font-size:3em;color:black;"> Session 4 - Lab Exercise 6</p>

**Group 6**
1. Ahmed Ali
2. Nichapat Boonprasertsri
3. Anuphap Chansatit
4. Karthikeyan Jeyabalasuntharam
5. Yat Chit Law
6. Halari Shanpru
7. Vitchaya Siripoppohn
8. Chotiros Srisiam

In [2]:
# you can download the original Kyphosis dataset from kaggle at https://www.kaggle.com/abbasit/kyphosis-dataset?select=kyphosis.csv

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [3]:
Kyphosis_df = pd.read_csv('kyphosis.csv')

In [4]:
from sklearn.model_selection import train_test_split

In [5]:
X = Kyphosis_df.drop('Kyphosis',axis=1)
y = Kyphosis_df['Kyphosis']

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

In [7]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
# selecting the right "parameters/hyperparameters of ML models can significantly affect their performacnce".
# let's begin exploring those parameters more for a DT model.
'''
DecisionTreeClassifier(
    criterion='gini',
    splitter='best',
    max_depth=None,
    min_samples_split=2,
    min_samples_leaf=1,
    min_weight_fraction_leaf=0.0,
    max_features=None,
    random_state=None,
    max_leaf_nodes=None,
    min_impurity_decrease=0.0,
    min_impurity_split=None,
    class_weight=None,
    presort=False,
)
'''
# Let's consider these three parameters: criterion, splitter, min_samples_split
    # for parameters consider two cases 'gini' and 'entropy'
    # for splitter consider two cases 'best' and 'random'
    # for min_samples_split consider two cases 2 and 5
    # explain and dicuss your findings

In [17]:
from sklearn.metrics import confusion_matrix

# Define a function to evaluate Decision Tree models with different parameters
def evaluate_decision_tree(criterion, splitter, min_samples_split):
    dt_model = DecisionTreeClassifier(
        criterion=criterion,
        splitter=splitter,
        min_samples_split=min_samples_split,
        random_state=42  # You can set a specific random_state for reproducibility
    )
    dt_model.fit(X_train, y_train)
    y_pred = dt_model.predict(X_test)

    print(f"Parameters: Criterion='{criterion}', Splitter='{splitter}', Min Samples Split={min_samples_split}")

    # Calculate confusion matrix using the imported function
    cm = confusion_matrix(y_test, y_pred)
    print("Confusion Matrix:")
    print(cm)

    # Calculate precision, recall, and accuracy manually
    tp = cm[1,1]
    fp = cm[0,1]
    fn = cm[1,0]
    tn = cm[0,0]
    precision = tp / (tp + fp)
    recall = tp / (tp + fn)
    accuracy = (tp + tn) / (tp + tn + fp + fn)

    print(f"Precision: {precision}")
    print(f"Recall: {recall}")
    print(f"Accuracy: {accuracy}")
    print("\n")

# Evaluate the Decision Tree models with different parameter combinations
criteria = ['gini', 'entropy']
splitters = ['best', 'random']
min_samples_splits = [2, 5]

for criterion in criteria:
    for splitter in splitters:
        for min_samples_split in min_samples_splits:
            evaluate_decision_tree(criterion, splitter, min_samples_split)

Parameters: Criterion='gini', Splitter='best', Min Samples Split=2
Confusion Matrix:
[[19  2]
 [ 3  1]]
Precision: 0.3333333333333333
Recall: 0.25
Accuracy: 0.8


Parameters: Criterion='gini', Splitter='best', Min Samples Split=5
Confusion Matrix:
[[21  0]
 [ 3  1]]
Precision: 1.0
Recall: 0.25
Accuracy: 0.88


Parameters: Criterion='gini', Splitter='random', Min Samples Split=2
Confusion Matrix:
[[19  2]
 [ 3  1]]
Precision: 0.3333333333333333
Recall: 0.25
Accuracy: 0.8


Parameters: Criterion='gini', Splitter='random', Min Samples Split=5
Confusion Matrix:
[[19  2]
 [ 3  1]]
Precision: 0.3333333333333333
Recall: 0.25
Accuracy: 0.8


Parameters: Criterion='entropy', Splitter='best', Min Samples Split=2
Confusion Matrix:
[[19  2]
 [ 3  1]]
Precision: 0.3333333333333333
Recall: 0.25
Accuracy: 0.8


Parameters: Criterion='entropy', Splitter='best', Min Samples Split=5
Confusion Matrix:
[[21  0]
 [ 3  1]]
Precision: 1.0
Recall: 0.25
Accuracy: 0.88


Parameters: Criterion='entropy', Splitte

# Findings:

1. **Criterion Choice (Gini vs. Entropy):** The choice of criterion (either 'gini' or 'entropy') does not significantly affect the model's performance. In all cases, the confusion matrices, precision, recall, and accuracy are nearly identical for the same combination of other parameters.

2. **Splitter Choice (Best vs. Random):** The choice of splitter ('best' or 'random') also does not have a substantial impact on model performance. For each combination of criterion and min_samples_split, the results are consistent regardless of the splitter choice.

3. **Min Samples Split (2 vs. 5):** Varying the minimum number of samples required to split a node (2 vs. 5) doesn't show a significant pattern in performance. The impact of this parameter appears to be minimal in this specific scenario.

4. **Accuracy and Model Evaluation:** The accuracy metric provides an overall measure of model performance. Across all tested parameter combinations, accuracy ranges from 0.8 to 0.88.

5. **Precision and Recall:** Precision measures the proportion of true positive predictions among all positive predictions, while recall measures the proportion of true positives among all actual positives. In all cases, precision is 0.333 (or 1.0 for some combinations), and recall is 0.25. These metrics suggest that the models have relatively low precision and recall values.